### DISCUSSION TITLE
How is Diesel safe?

### DISCUSSION THREAD

**Discussion by andrewbaxter** [diesel-rs/diesel#3466] [Q&A]
Safe can mean a lot of things, and compared to a super-typed language like Rust the minimum bound ORM safety of "no query syntax errors" isn't a whole lot of safety.  Diesel is "safe" but I couldn't find any information about what that actually means, in practice.

This is probably a simplification, but three levels of safety I care about are
1. Queries don't bail out on syntax errors
2. Runtime errors packing/unpacking query parameters and results (i.e. query returns 4 things, I try to access 5th, crash)
3. Types are guaranteed during 2, including newtypes, json schemas, etc.

I was using SQLx which gets you 1 and 2 but not 3 and I ended spending a lot of time dealing with issues due to the lack of 3.

**Comment by weiznich** [comment]
This question is really hard to answer as it is quite unspecific. I will try my best and add some general statements here. Please consider asking for clarification if something remains unclear.

"Safe" in diesels documentation mostly refers to the fact that diesel catches a lot of issues at compile time that would be otherwise runtime errors. This includes the following issues:

* Any kind of type mismatch between what's in the database and what's in your code
* Constructing queries and query parts that are just invalid, by for example disallowing columns from other tables (that are not joined), enforcing group by rules, …


It does not include things like handling out of bounds access. This will result in an out of bounds error as is normal rust. That's because diesel normally returns all values as `Vec<_>` and indexing outside of the length of the vector is obviously a out of bounds issue. 

I think that covers what you listed as point 1 and 2.

I'm not exactly sure what you expect by the third point. Diesel uses a strict type-system for all of that. You also can add your own types, including an newtype wrappers. This allows to configure almost anything.

For json-schemas: We do support them, but they are by nature dynamically typed, which means you cannot do any compile time checking there other than: This is a json field, so the return type is json.

**Reply by andrewbaxter** [reply]
So for example,
```
struct UserId(i32);
struct AccountId(i32);

#[derive(Queryable)]
#[diesel(table_name=posts)]
pub struct PostSummary {
    pub id: UserId,
    pub title: String,
}

#[derive(Queryable)]
#[diesel(table_name=posts)]
pub struct FullPost {
    pub id: AccountId,
    pub title: String,
    pub body: String,
    pub attachments: HashMap<String, Vec<u8>>,
}
```

Would it catch that the id type is different in the two queries?  Looking at the examples I assume Diesel is limited in type checking to just the SQL types.

**Reply by weiznich** [reply]
Such a pattern is also possible. You would need to implement some traits for the new type wrappers, but otherwise that should work fine. There is also an [extension crate](https://crates.io/crates/diesel-derive-newtype) for this.

**Reply by andrewbaxter** [reply]
Ah thanks! I think you read my message backwards, but that link confirms that Diesel is _unsafe_ in this regard which is what I wanted to know.  Quoting the extension crate

> even though your newtypes can be used anywhere the underlying type can be used, the underlying types, or any other newtypes of the same underlying type, can be used as well.

So Diesel doesn't provide type safety beyond the built in SQL types, and the newtype extension doesn't provide any additional safety in this regard (aside from serialization/deserialization safety).

Thanks again!

**Reply by weiznich** [reply]
That's what's implemented by the extension crate, which is not everything you could do there. They have a pretty good "Limitations" section in their readme, that outlines the various trade-offs for their solution. They also outline what needs to be done to get your desired way. To be clear, that's something that is totally possible with diesel, it is just not the built-in default as I consider that to be to brittle to be used.

To give a more or less complete example of what you are looking for:

```rust
table! {
    use crate::sql_types::UserIdSqlType;
    use diesel::sql_types::*;

    users {
        id -> UserIdSqlType,
       name -> Text,
    }
}

mod sql_types {
    #[derive(Clone, Copy, diesel::sql_types::SqlType)]
    #[diesel(postgres_type(oid = 23, array_oid = 1007))]
    pub struct UserIdSqlType;
}

#[derive(diesel::deserialize::FromSqlRow, diesel::expression::AsExpression)]
#[diesel(sql_type = UserIdSqlType)]
struct UserId(i32);

impl FromSql<crate::sql_types::UserIdSqlType, diesel::pg::Pg> for UserId {
     fn from_sql(bytes: diesel::backend::RawValue<'_, diesel::pg::Pg>) -> diesel::deserialize::Result<Self> {
          FromSql::<diesel::sql_types::Integer, diesel::pg::Pg>::from_sql(bytes).map(Self)
     }
}

impl ToSql<crate::sql_types::UserIdSqlType, diesel::pg::Pg> for UserId {
    fn to_sql<'b>(&'b self, out: &mut diesel::deserialize::Output<'b, '_, DB>) -> diesel::deserialize::Result {
        <_ as ToSql<diesel::sql_types::Integer, diesel::pg::Pg>>::to_sql(&self.0)
    }
}
```

Using this implementation it's impossible to mix up any types anywhere, as long as you don't reuse the `UserIdSqlType` for other columns. All of that can be implemented by a third party crate and likely simplified by using macros.

**Reply by andrewbaxter** [reply]
Oh nice, that worked!
```
use diesel::{
    deserialize::FromSql, serialize::ToSql, table, Connection, Insertable, PgConnection,
    RunQueryDsl,
};
use sql_types::{AccountIdSqlType, UserIdSqlType};

table! {
    use crate::sql_types::UserIdSqlType;
    use diesel::sql_types::*;

    users {
        id -> UserIdSqlType,
       name -> Text,
    }
}

mod sql_types {
    #[derive(Clone, Copy, diesel::sql_types::SqlType)]
    #[diesel(postgres_type(oid = 23, array_oid = 1007))]
    pub struct UserIdSqlType;
    #[derive(Clone, Copy, diesel::sql_types::SqlType)]
    #[diesel(postgres_type(oid = 23, array_oid = 1007))]
    pub struct AccountIdSqlType;
}

#[derive(diesel::deserialize::FromSqlRow, diesel::expression::AsExpression, Debug)]
#[diesel(sql_type = UserIdSqlType)]
struct UserId(i32);

impl FromSql<crate::sql_types::UserIdSqlType, diesel::pg::Pg> for UserId {
    fn from_sql(
        bytes: diesel::backend::RawValue<'_, diesel::pg::Pg>,
    ) -> diesel::deserialize::Result<Self> {
        FromSql::<diesel::sql_types::Integer, diesel::pg::Pg>::from_sql(bytes).map(Self)
    }
}

impl ToSql<crate::sql_types::UserIdSqlType, diesel::pg::Pg> for UserId {
    fn to_sql<'b>(
        &'b self,
        out: &mut diesel::serialize::Output<'b, '_, diesel::pg::Pg>,
    ) -> diesel::serialize::Result {
        <_ as ToSql<diesel::sql_types::Integer, diesel::pg::Pg>>::to_sql(&self.0, out)
    }
}

#[derive(diesel::deserialize::FromSqlRow, diesel::expression::AsExpression, Debug)]
#[diesel(sql_type = AccountIdSqlType)]
struct AccountId(i32);

impl FromSql<crate::sql_types::AccountIdSqlType, diesel::pg::Pg> for AccountId {
    fn from_sql(
        bytes: diesel::backend::RawValue<'_, diesel::pg::Pg>,
    ) -> diesel::deserialize::Result<Self> {
        FromSql::<diesel::sql_types::Integer, diesel::pg::Pg>::from_sql(bytes).map(Self)
    }
}

impl ToSql<crate::sql_types::AccountIdSqlType, diesel::pg::Pg> for AccountId {
    fn to_sql<'b>(
        &'b self,
        out: &mut diesel::serialize::Output<'b, '_, diesel::pg::Pg>,
    ) -> diesel::serialize::Result {
        <_ as ToSql<diesel::sql_types::Integer, diesel::pg::Pg>>::to_sql(&self.0, out)
    }
}

#[derive(Insertable)]
#[diesel(table_name=users)]
struct NewUser<'a> {
    id: &'a AccountId,
    name: &'a str,
}

fn main() {
    let database_url = "postgres://postgres:myscretpassword@localhost/testing";
    let mut conn = PgConnection::establish(&database_url).unwrap();
    diesel::insert_into(users::table)
        .values(&NewUser {
            id: &AccountId(999),
            name: "x",
        })
        .execute(&mut conn)
        .unwrap();
}
```
fails to compile with the error 
```
the trait bound `AccountId: diesel::Expression` is not satisfied
...
required because of the requirements on the impl of `diesel::Expression` for `&'a AccountId`
required because of the requirements on the impl of `AsExpression<UserIdSqlType>` for `&'a
```

**Reply by andrewbaxter** [reply]
How does this work with migrations? I imagine the cli re-writes the schema with sql types instead of custom types.  Or is that what you referring to when you mentioned it being fragile?

**Reply by weiznich** [reply]
> How does this work with migrations? I imagine the cli re-writes the schema with sql types instead of custom types.

That is something that needs to be handled outside of diesel. For example by using a patch file or by generating your own variant of a schema file. (That is quite easy)

>  Or is that what you referring to when you mentioned it being fragile?

That was mostly referring to the fact that this restricts what's possible to express with the query builder. We do generally aim for allowing whatever would be syntactically valid SQL. Such an approach disallows certain pattern that would be valid SQL, like for example `SELECT user.id + post.id FROM users INNER JOIN posts …` (or any other more general operation between two values of the same database type). That is fine for specific use-cases, but nothing we want to assume for the general use-case of diesel.
